lib/remote_node_manager.go

```diff
@@ -0,0 +1,297 @@
+	indexer.GetAllRemoteNodes().RemoveRemoteNode(rn)
+	indexer.GetValidatorIndex().RemoveRemoteNode(rn)
+	indexer.GetNonValidatorOutboundIndex().RemoveRemoteNode(rn)
+	indexer.GetNonValidatorInboundIndex().RemoveRemoteNode(rn)
+}
+
+func (manager *RemoteNodeManager) newRemoteNode() *RemoteNode {
+	id := atomic.AddUint64(&manager.remoteNodeIndex, 1)
+	remoteNodeId := NewRemoteNodeId(id)
+	latestBlockHeight := uint64(manager.bc.BlockTip().Height)
+	return NewRemoteNode(remoteNodeId, manager.srv, manager.cmgr, manager.keystore, manager.params,
+		manager.minTxFeeRateNanosPerKB, latestBlockHeight, manager.nodeServices)
+}
+
+// FIXME
+func (manager *RemoteNodeManager) SendMessageToPeer(peer *Peer, desoMessage DeSoMessage) {
+	rn := manager.GetRemoteNodeFromPeer(peer)
+	if rn == nil {
+		return
+	}
+
+	rn.SendMessage(desoMessage)
+}
+
+func (manager *RemoteNodeManager) CreateValidatorConnection(netAddr *wire.NetAddress, publicKey bls.SerializedPublicKey) error {
+	if netAddr == nil {
+		return fmt.Errorf("RemoteNodeManager.CreateValidatorConnection: netAddr is nil")
+	}
+
+	remoteNode := manager.newRemoteNode()
+	if err := remoteNode.DialPersistentOutboundConnection(netAddr); err != nil {
+		return errors.Wrapf(err, "RemoteNodeManager.CreateValidatorConnection: Problem calling DialPersistentOutboundConnection "+
+			"for addr: (%s:%v)", netAddr.IP.String(), netAddr.Port)
+	}
+	manager.SetRemoteNode(remoteNode)
+	manager.GetRemoteNodeIndexer().GetValidatorIndex().Add(publicKey, remoteNode)
+	return nil
+}
+
+func (manager *RemoteNodeManager) CreatePersistentOutboundConnectionNetAddress(netAddr *wire.NetAddress) error {
+	if netAddr == nil {
+		return fmt.Errorf("RemoteNodeManager.CreatePersistentOutboundConnectionNetAddress: netAddr is nil")
+	}
+
+	remoteNode := manager.newRemoteNode()
+	if err := remoteNode.DialPersistentOutboundConnection(netAddr); err != nil {
+		return errors.Wrapf(err, "RemoteNodeManager.CreatePersistentOutboundConnectionNetAddress: Problem calling DialPersistentOutboundConnection "+
+			"for addr: (%s:%v)", netAddr.IP.String(), netAddr.Port)
+	}
+	manager.SetRemoteNode(remoteNode)
+	manager.GetRemoteNodeIndexer().GetNonValidatorOutboundIndex().Add(remoteNode.GetId(), remoteNode)
+	return nil
+}
+
+func (manager *RemoteNodeManager) CreateOutboundConnectionNetAddress(netAddr *wire.NetAddress) (*RemoteNode, error) {
+	if netAddr == nil {
+		return nil, fmt.Errorf("RemoteNodeManager.CreateOutboundConnectionNetAddress: netAddr is nil")
+	}
+
+	remoteNode := manager.newRemoteNode()
+	if err := remoteNode.DialOutboundConnection(netAddr); err != nil {
+		return nil, errors.Wrapf(err, "RemoteNodeManager.CreateOutboundConnectionNetAddress: Problem calling DialOutboundConnection "+
+			"for addr: (%s:%v)", netAddr.IP.String(), netAddr.Port)
+	}
+	manager.SetRemoteNode(remoteNode)
+	manager.GetRemoteNodeIndexer().GetNonValidatorOutboundIndex().Add(remoteNode.GetId(), remoteNode)
+	return remoteNode, nil
```

Comment:
I notice you don't return the remoteNode in the `CreatePersistentOutboundConnectionNetAddress()` function above. Any reason to return it here and not there? Just want to make sure there's no bug